apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,235 @@
+  CallHandler,
+  Logger,
+  ConflictException,
+  HttpException,
+  InternalServerErrorException,
+} from '@nestjs/common';
+import { CacheService } from '@novu/application-generic';
+import { Observable, of, throwError } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+import { createHash } from 'crypto';
+import * as jwt from 'jsonwebtoken';
+import { IJwtPayload } from '@novu/shared';
+
+const LOG_CONTEXT = 'IdempotencyInterceptor';
+const IDEMPOTENCY_CACHE_TTL = 60 * 60 * 6; //6h
+const IDEMPOTENCY_PROGRESS_TTL = 60 * 5; //5min
+
+const HEADER_KEYS = {
+  IDEMPOTENCY: 'idempotency-key',
+  RETRY_AFTER: 'retry-after',
+  IDEMPOTENCY_CONFLICT: 'x-idempotency-conflict',
+};
+
+enum ReqStatusEnum {
+  PROGRESS = 'in-progress',
+  SUCCESS = 'success',
+  ERROR = 'error',
+}
+
+@Injectable()
+export class IdempotencyInterceptor implements NestInterceptor {
+  constructor(private readonly cacheService: CacheService) {}
+
+  async intercept(context: ExecutionContext, next: CallHandler): Promise<Observable<any>> {
+    const request = context.switchToHttp().getRequest();
+    const idempotencyKey = this.getIdempotencyKey(context);
+    const isEnabled = process.env.API_IDEMPOTENCY_ENABLED == 'true';
+    if (!isEnabled || !idempotencyKey || !['post', 'patch'].includes(request.method.toLowerCase())) {
+      return next.handle();
+    }
+    const cacheKey = this.getCacheKey(context);
+
+    try {
+      const bodyHash = this.hashRequestBody(request.body);
+      //if 1st time we are seeing the request, marks the request as in-progress if not, does nothing
+      const isNewReq = await this.setCache(
+        cacheKey,
+        { status: ReqStatusEnum.PROGRESS, bodyHash },
+        IDEMPOTENCY_PROGRESS_TTL,
+        true
+      );
+      // Check if the idempotency key is in the cache
+      if (isNewReq) {
+        return await this.handleNewRequest(context, next, bodyHash);
+      } else {
+        return await this.handlerDuplicateRequest(context, bodyHash);
+      }
+    } catch (err) {
+      Logger.warn(
+        `An error occurred while making idempotency check, proceeding without idempotency. key:${idempotencyKey}. error: ${err.message}`,
+        LOG_CONTEXT
+      );
+    }
+
+    //something unexpected happened, both cached response and handler did not execute as expected,so proceed w/o idempotency
+    return next.handle();
+  }
+
+  private getIdempotencyKey(context: ExecutionContext): string | undefined {
+    const request = context.switchToHttp().getRequest();
+
+    return request.headers[HEADER_KEYS.IDEMPOTENCY];
+  }
+
+  private getReqUser(context: ExecutionContext): IJwtPayload | null {
+    const req = context.switchToHttp().getRequest();
+    if (req?.user?.organizationId) {
+      return req.user;
+    }
+    if (req.headers?.authorization?.length) {
+      const token = req.headers.authorization.split(' ')[1];
+      if (token) {
+        return jwt.decode(token);
+      }
+    }
+
+    return null;
```

Comment:
We need to test and update this function to return a defined `organizationId` when API Key authentication is used.

In NestJS, interceptors execute after guards in the [request lifecycle](https://docs.nestjs.com/faq/request-lifecycle), so we should be able to access a decoded user in this interceptor.

Reply:
yes auth gaurds are executed before but, when api key is used we are not attaching the decoded user to the request, its only attached when using bearer token. thats why i added this change, we access the user in route method using custom param decorator

Reply:
Whilst `jwt.decode` is a fairly low cost operation, running it on every request could significantly increase memory consumption, especially if the JWT payload grows. Furthermore we are in fact signing a new token with the latency cost of encryption every time we authenticate an API key.

I think we should move ahead with this implementation for now and address this issue separately.

To future-proof this implementation from a performance point of view, I'd suggest we update the Auth guard to return the user when an API token is present in a similar way to the [JwtAuthGuard built into @nestjs/passport](see https://docs.nestjs.com/security/authentication#implementing-the-authentication-guard), as it's an anti-pattern to encrypt/sign a JWT for further downstream use inside an Auth guard.

@LetItRock @Cliftonz, we should consider this to be tech debt and action it as so. Let me know if you agree and I can raise a ticket.